client/state/current-user/gravatar-status/actions.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * External dependencies
+ */
+import request from 'superagent';
+
+import {
+	GRAVATAR_UPLOAD_RECEIVE,
+	GRAVATAR_UPLOAD_REQUEST,
+	GRAVATAR_UPLOAD_REQUEST_SUCCESS,
+	GRAVATAR_UPLOAD_REQUEST_FAILURE
+} from 'state/action-types';
+import {
+	GRAVATAR_CACHE_EXPIRATION
+} from 'state/current-user/gravatar-status/constants';
+
+export function uploadGravatar( file, bearerToken, email ) {
+	return dispatch => {
+		dispatch( { type: GRAVATAR_UPLOAD_REQUEST } );
+		const data = new window.FormData();
+		data.append( 'filedata', file );
+		data.append( 'account', email );
+		return request
+			.post( 'https://api.gravatar.com/v1/upload-image' )
+			.send( data )
+			.set( 'Authorization', 'Bearer ' + bearerToken )
+			.then( () => {
+				const fileReader = new window.FileReader( file );
+				fileReader.addEventListener( 'load', function() {
+					dispatch( {
+						type: GRAVATAR_UPLOAD_RECEIVE,
+						expiration: Date.now() + GRAVATAR_CACHE_EXPIRATION,
+						tempImage: fileReader.result
+					} );
+					dispatch( {
+						type: GRAVATAR_UPLOAD_REQUEST_SUCCESS
+					} );
+				} );
+				fileReader.readAsDataURL( file );
```

Comment:
I didn't add an error handler for FileReader since the error is caught by the next `catch` block in the promise chain. Right now it displays only the default error notice, and my plan is to add more specific error messages for error responses from Gravatar.com.

I wasn't going to add a more specific error message for any FileReader errors, but you're right it would be better to add something in since the Gravatar will have been uploaded fine, but we won't be able to display it right away.

Ooh, and I just added a change for showing a cache-busted URL as a bug fix for Chrome, so it's possible to re-use that for showing the new image even if the FileReader throws an error. It's an edgecase issue, so I probably won't get to it on the first iteration, but it would be pretty simple to add in given the actions we have so far 😄

FileReader errors should be rare since the image is pre-vetted by both Gravatar.com and also error checking in `EditGravatar` (not yet implemented).